frontend/src/util/activity.js

```diff
@@ -0,0 +1,191 @@
+    reactionList: [],
+    commentList: commentList,
+    newestMessage: first,
+    content: content
+  }
+}
+
+const getActivityId = (message) => {
+  const [entityType, , subEntityType] = message.event_type.split('.')
+  let id = null
+  switch (entityType) {
+    case TLM_ET.CONTENT:
+    case TLM_ET.MENTION:
+      id = (subEntityType === TLM_ST.COMMENT)
+        ? message.fields.content.parent_id
+        : message.fields.content.content_id
+      break
+    case TLM_ET.SHAREDSPACE_MEMBER:
+    case TLM_ET.SHAREDSPACE_SUBSCRIPTION:
+      id = `e${message.event_id}`
+      break
+  }
+  if (id === null) return null
+  return `${entityType}-${id}`
+}
+
+const createActivity = async (activityId, activityMessageList, apiUrl) => {
+  if (activityId.startsWith(TLM_ET.CONTENT)) {
+    return await createContentActivity(activityId, activityMessageList, apiUrl)
+  } else if (activityId.startsWith(TLM_ET.SHAREDSPACE_MEMBER)) {
+    return createMemberActivity(activityId, activityMessageList)
+  } else if (activityId.startsWith(TLM_ET.SHAREDSPACE_SUBSCRIPTION)) {
+    return createSubscriptionActivity(activityId, activityMessageList)
+  }
+}
+
+const groupMessageListByActivityId = (messageList) => {
+  const activityMap = new Map()
+  for (const message of messageList) {
+    const activityId = getActivityId(message)
+    if (activityId !== null) {
+      const activityMessageList = activityMap.get(activityId) || []
+      activityMap.set(activityId, [...activityMessageList, message])
+    }
+  }
+
+  return activityMap
+}
+
+const createActivityListFromActivityMap = async (activityMap, apiUrl) => {
+  const activityCreationList = []
+  for (const [activityId, activityMessageList] of activityMap) {
+    activityCreationList.push(createActivity(activityId, activityMessageList, apiUrl))
+  }
+  return await Promise.all(activityCreationList)
+}
+
+/**
+ * Merge an activity list with message/TLM list
+ * messages are assumed to be older.
+ * Activities are returned in newest to oldest order.
+ * INFO - SG - 2020-11-12 - this function assumes that the message list is already ordered from newest to oldest.
+ */
+export const mergeWithActivityList = async (messageList, activityList, apiUrl) => {
+  const activityMap = groupMessageListByActivityId(messageList)
+
+  for (const activity of activityList) {
+    if (activityMap.has(activity.id)) {
+      activityMap.delete(activity.id)
+    }
+  }
+
+  const newActivityList = await createActivityListFromActivityMap(activityMap, apiUrl)
+  return [...activityList, ...newActivityList]
+}
+
+/**
+ * Sort an activity list from newest to oldest
+ * Uses the newestMessage event_id attribute as the sort key.
+ */
+export const sortActivityList = (activityList) => {
+  return [...activityList].sort((a, b) => b.newestMessage.event_id - a.newestMessage.event_id)
+}
+
+const updateActivity = (message, activity) => {
+  const isComment = message.event_type.endsWith(`.${TLM_ST.COMMENT}`)
```

Comment:
Should you do the same thing for mention?

Reply:
No this is OK, the exception for comments is to avoid replacing the "real" content by the comment. In a mention TLM the `content` field is the real content.
But I agree this is overly complicated, we possibly should have `comment.xxx` tlms rather than those `content.xxx.comment` which are not very useful and required many tweaks (adding parent_content_type, parent_label and many cases in the frontend code).